TestStorePoolFindDeadReplicas is flaky

The following test appears to have failed:

[#9685](https://circleci.com/gh/cockroachdb/cockroach/9685):

```
--- PASS: TestStorePoolGetStoreList (0.01s)
=== RUN   TestStorePoolGetStoreDetails
--- PASS: TestStorePoolGetStoreDetails (0.01s)
=== RUN   TestStorePoolFindDeadReplicas
W1130 19:16:15.683799 972 storage/store_pool.go:58  store 1 on node 1 is now considered offline
--- FAIL: TestStorePoolFindDeadReplicas (0.01s)
	store_pool_test.go:357: expected no dead replicas initially, found 1 ([{1 1 1}])
=== RUN   TestStoreInitAndBootstrap
I1130 19:16:15.687243 972 multiraft/multiraft.go:557  node 1 starting
I1130 19:16:15.687730 972 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestStoreInitAndBootstrap (0.01s)
=== RUN   TestBootstrapOfNonEmptyStore
I1130 19:16:15.697278 972 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBootstrapOfNonEmptyStore (0.00s)
=== RUN   TestStoreAddRemoveRanges
I1130 19:16:15.705894 972 multiraft/multiraft.go:557  node 1 starting
--
I1130 19:16:44.697772 972 storage/replica.go:1369  gossiping system config from store 1, range 2
I1130 19:16:44.711028 972 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestStoreSplitReadRace (0.05s)
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.48s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	31.859s
=== RUN   TestBatchBasics
I1130 19:16:13.452760 977 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN   TestBatchGet
I1130 19:16:13.458178 977 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I1130 19:16:13.465316 977 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/server/server.go
+++ b/server/server.go
@@ -119,7 +119,7 @@ func NewServer(ctx *Context, stopper *stop.Stopper) (*Server, error) {
 	s.rpc = rpc.NewServer(util.MakeUnresolvedAddr("tcp", addr), rpcContext)
 	s.stopper.AddCloser(s.rpc)
 	s.gossip = gossip.New(rpcContext, s.ctx.GossipBootstrapResolvers)
-	s.storePool = storage.NewStorePool(s.gossip, ctx.TimeUntilStoreDead, stopper)
+	s.storePool = storage.NewStorePool(s.gossip, s.clock, ctx.TimeUntilStoreDead, stopper)
 
 	feed := util.NewFeed(stopper)
 	tracer := tracer.NewTracer(feed, addr)

--- a/storage/allocator_test.go
+++ b/storage/allocator_test.go
@@ -165,11 +165,12 @@ var multiDCStores = []*roachpb.StoreDescriptor{
 // use in tests. Stopper must be stopped by the caller.
 func createTestAllocator() (*stop.Stopper, *gossip.Gossip, *StorePool, Allocator) {
 	stopper := stop.NewStopper()
-	rpcContext := rpc.NewContext(&base.Context{}, hlc.NewClock(hlc.UnixNano), stopper)
+	clock := hlc.NewClock(hlc.UnixNano)
+	rpcContext := rpc.NewContext(&base.Context{}, clock, stopper)
 	g := gossip.New(rpcContext, gossip.TestBootstrap)
 	// Have to call g.SetNodeID before call g.AddInfo
 	g.SetNodeID(roachpb.NodeID(1))
-	storePool := NewStorePool(g, TestTimeUntilStoreDeadOff, stopper)
+	storePool := NewStorePool(g, clock, TestTimeUntilStoreDeadOff, stopper)
 	a := MakeAllocator(storePool, AllocatorOptions{AllowRebalance: true})
 	return stopper, g, storePool, a
 }
@@ -1083,7 +1084,7 @@ func Example_rebalancing() {
 	g.SetNodeID(roachpb.NodeID(1))
 	stopper := stop.NewStopper()
 	defer stopper.Stop()
-	sp := NewStorePool(g, TestTimeUntilStoreDeadOff, stopper)
+	sp := NewStorePool(g, hlc.NewClock(hlc.UnixNano), TestTimeUntilStoreDeadOff, stopper)
 	alloc := MakeAllocator(sp, AllocatorOptions{AllowRebalance: true, Deterministic: true})
 
 	var wg sync.WaitGroup

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -1179,7 +1179,8 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 	// This can't use SucceedsWithin as using the backoff mechanic won't work
 	// as it requires a specific cadence of re-gossiping the alive stores to
 	// maintain their alive status.
-	ticker := time.NewTicker(storage.TestTimeUntilStoreDead / 2)
+	tickerDur := storage.TestTimeUntilStoreDead / 2
+	ticker := time.NewTicker(tickerDur)
 	defer ticker.Stop()
 
 	maxTime := 5 * time.Second
@@ -1190,6 +1191,8 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 		case <-maxTimeout:
 			t.Fatalf("Failed to remove the dead replica within %s", maxTime)
 		case <-ticker.C:
+			mtc.manualClock.Increment(int64(tickerDur))
+
 			// Keep gossiping the alive stores.
 			sg.GossipWithFunction(aliveStoreIDs, func() {
 				mtc.stores[0].GossipStore()

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -214,7 +214,7 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
 		if m.timeUntilStoreDead == 0 {
 			m.timeUntilStoreDead = storage.TestTimeUntilStoreDeadOff
 		}
-		m.storePool = storage.NewStorePool(m.gossip, m.timeUntilStoreDead, m.clientStopper)
+		m.storePool = storage.NewStorePool(m.gossip, m.clock, m.timeUntilStoreDead, m.clientStopper)
 	}
 
 	// Always create the first sender.

--- a/storage/simulation/cluster.go
+++ b/storage/simulation/cluster.go
@@ -64,7 +64,7 @@ func createCluster(stopper *stop.Stopper, nodeCount int, epochWriter, actionWrit
 	clock := hlc.NewClock(hlc.UnixNano)
 	rpcContext := rpc.NewContext(&base.Context{}, clock, stopper)
 	g := gossip.New(rpcContext, gossip.TestBootstrap)
-	storePool := storage.NewStorePool(g, storage.TestTimeUntilStoreDeadOff, stopper)
+	storePool := storage.NewStorePool(g, clock, storage.TestTimeUntilStoreDeadOff, stopper)
 	c := &Cluster{
 		stopper:   stopper,
 		clock:     clock,

--- a/storage/store_pool.go
+++ b/storage/store_pool.go
@@ -25,6 +25,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/gossip"
 	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/stop"
 )
@@ -44,13 +45,13 @@ type storeDetail struct {
 	dead            bool
 	gossiped        bool // Was this store updated via gossip?
 	timesDied       int
-	foundDeadOn     time.Time
-	lastUpdatedTime time.Time // This is also the priority for the queue.
-	index           int       // index of the item in the heap, required for heap.Interface
+	foundDeadOn     roachpb.Timestamp
+	lastUpdatedTime roachpb.Timestamp // This is also the priority for the queue.
+	index           int               // index of the item in the heap, required for heap.Interface
 }
 
 // markDead sets the storeDetail to dead(inactive).
-func (sd *storeDetail) markDead(foundDeadOn time.Time) {
+func (sd *storeDetail) markDead(foundDeadOn roachpb.Timestamp) {
 	sd.dead = true
 	sd.foundDeadOn = foundDeadOn
 	sd.timesDied++
@@ -59,7 +60,7 @@ func (sd *storeDetail) markDead(foundDeadOn time.Time) {
 
 // markAlive sets the storeDetail to alive(active) and saves the updated time
 // and descriptor.
-func (sd *storeDetail) markAlive(foundAliveOn time.Time, storeDesc roachpb.StoreDescriptor, gossiped bool) {
+func (sd *storeDetail) markAlive(foundAliveOn roachpb.Timestamp, storeDesc roachpb.StoreDescriptor, gossiped bool) {
 	sd.desc = storeDesc
 	sd.dead = false
 	sd.gossiped = gossiped
@@ -77,7 +78,7 @@ func (pq storePoolPQ) Len() int {
 
 // Less implements the sort.Interface.
 func (pq storePoolPQ) Less(i, j int) bool {
-	return pq[i].lastUpdatedTime.Before(pq[j].lastUpdatedTime)
+	return pq[i].lastUpdatedTime.Less(pq[j].lastUpdatedTime)
 }
 
 // Swap implements the sort.Interface.
@@ -133,6 +134,7 @@ func (pq *storePoolPQ) dequeue() *storeDetail {
 // StorePool maintains a list of all known stores in the cluster and
 // information on their health.
 type StorePool struct {
+	clock              *hlc.Clock
 	timeUntilStoreDead time.Duration
 
 	// Each storeDetail is contained in both a map and a priorityQueue; pointers
@@ -144,8 +146,9 @@ type StorePool struct {
 
 // NewStorePool creates a StorePool and registers the store updating callback
 // with gossip.
-func NewStorePool(g *gossip.Gossip, timeUntilStoreDead time.Duration, stopper *stop.Stopper) *StorePool {
+func NewStorePool(g *gossip.Gossip, clock *hlc.Clock, timeUntilStoreDead time.Duration, stopper *stop.Stopper) *StorePool {
 	sp := &StorePool{
+		clock:              clock,
 		timeUntilStoreDead: timeUntilStoreDead,
 		stores:             make(map[roachpb.StoreID]*storeDetail),
 	}
@@ -176,7 +179,7 @@ func (sp *StorePool) storeGossipUpdate(_ string, content roachpb.Value) {
 		detail = &storeDetail{index: -1}
 		sp.stores[storeDesc.StoreID] = detail
 	}
-	detail.markAlive(time.Now(), storeDesc, true)
+	detail.markAlive(sp.clock.Now(), storeDesc, true)
 	sp.queue.enqueue(detail)
 }
 
@@ -193,9 +196,9 @@ func (sp *StorePool) start(stopper *stop.Stopper) {
 				timeout = sp.timeUntilStoreDead
 			} else {
 				// Check to see if the store should be marked as dead.
-				deadAsOf := detail.lastUpdatedTime.Add(sp.timeUntilStoreDead)
-				now := time.Now()
-				if now.After(deadAsOf) {
+				deadAsOf := detail.lastUpdatedTime.GoTime().Add(sp.timeUntilStoreDead)
+				now := sp.clock.Now()
+				if now.GoTime().After(deadAsOf) {
 					deadDetail := sp.queue.dequeue()
 					deadDetail.markDead(now)
 					// The next store might be dead as well, set the timeout to
@@ -204,7 +207,7 @@ func (sp *StorePool) start(stopper *stop.Stopper) {
 				} else {
 					// Store is still alive, schedule the next check for when
 					// it should timeout.
-					timeout = deadAsOf.Sub(now)
+					timeout = deadAsOf.Sub(now.GoTime())
 				}
 			}
 			sp.mu.Unlock()
@@ -229,7 +232,7 @@ func (sp *StorePool) getStoreDetail(storeID roachpb.StoreID) storeDetail {
 		// considered dead.
 		detail = &storeDetail{index: -1}
 		sp.stores[storeID] = detail
-		detail.markAlive(time.Now(), roachpb.StoreDescriptor{StoreID: storeID}, false)
+		detail.markAlive(sp.clock.Now(), roachpb.StoreDescriptor{StoreID: storeID}, false)
 		sp.queue.enqueue(detail)
 	}
 

--- a/storage/store_pool.go
+++ b/storage/store_pool.go
@@ -25,7 +25,6 @@ import (
 
 	"github.com/cockroachdb/cockroach/gossip"
 	"github.com/cockroachdb/cockroach/roachpb"
-	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/stop"
 )
@@ -134,8 +133,6 @@ func (pq *storePoolPQ) dequeue() *storeDetail {
 // StorePool maintains a list of all known stores in the cluster and
 // information on their health.
 type StorePool struct {
-	gossip             *gossip.Gossip
-	clock              *hlc.Clock
 	timeUntilStoreDead time.Duration
 
 	// Each storeDetail is contained in both a map and a priorityQueue; pointers

--- a/storage/store_pool_test.go
+++ b/storage/store_pool_test.go
@@ -53,21 +53,23 @@ var uniqueStore = []*roachpb.StoreDescriptor{
 
 // createTestStorePool creates a stopper, gossip and storePool for use in
 // tests. Stopper must be stopped by the caller.
-func createTestStorePool(timeUntilStoreDead time.Duration) (*stop.Stopper, *gossip.Gossip, *StorePool) {
+func createTestStorePool(timeUntilStoreDead time.Duration) (*stop.Stopper, *gossip.Gossip, *hlc.ManualClock, *StorePool) {
 	stopper := stop.NewStopper()
-	rpcContext := rpc.NewContext(&base.Context{}, hlc.NewClock(hlc.UnixNano), stopper)
+	mc := hlc.NewManualClock(0)
+	clock := hlc.NewClock(mc.UnixNano)
+	rpcContext := rpc.NewContext(&base.Context{}, clock, stopper)
 	g := gossip.New(rpcContext, gossip.TestBootstrap)
 	// Have to call g.SetNodeID before call g.AddInfo
 	g.SetNodeID(roachpb.NodeID(1))
-	storePool := NewStorePool(g, timeUntilStoreDead, stopper)
-	return stopper, g, storePool
+	storePool := NewStorePool(g, clock, timeUntilStoreDead, stopper)
+	return stopper, g, mc, storePool
 }
 
 // TestStorePoolGossipUpdate ensures that the gossip callback in StorePool
 // correctly updates a store's details.
 func TestStorePoolGossipUpdate(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	stopper, g, sp := createTestStorePool(TestTimeUntilStoreDead)
+	stopper, g, _, sp := createTestStorePool(TestTimeUntilStoreDead)
 	defer stopper.Stop()
 	sg := gossiputil.NewStoreGossiper(g)
 
@@ -90,8 +92,13 @@ func TestStorePoolGossipUpdate(t *testing.T) {
 }
 
 // waitUntilDead will block until the specified store is marked as dead.
-func waitUntilDead(t *testing.T, sp *StorePool, storeID roachpb.StoreID) {
+func waitUntilDead(t *testing.T, mc *hlc.ManualClock, sp *StorePool, storeID roachpb.StoreID) {
+	lastTime := time.Now()
 	util.SucceedsWithin(t, 10*TestTimeUntilStoreDead, func() error {
+		curTime := time.Now()
+		mc.Increment(curTime.UnixNano() - lastTime.UnixNano())
+		lastTime = curTime
+
 		sp.mu.RLock()
 		defer sp.mu.RUnlock()
 		store, ok := sp.stores[storeID]
@@ -111,7 +118,7 @@ func waitUntilDead(t *testing.T, sp *StorePool, storeID roachpb.StoreID) {
 // times out and that it will be revived after a new update is received.
 func TestStorePoolDies(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	stopper, g, sp := createTestStorePool(TestTimeUntilStoreDead)
+	stopper, g, mc, sp := createTestStorePool(TestTimeUntilStoreDead)
 	defer stopper.Stop()
 	sg := gossiputil.NewStoreGossiper(g)
 	sg.GossipStores(uniqueStore, t)
@@ -138,7 +145,7 @@ func TestStorePoolDies(t *testing.T) {
 	}
 
 	// Timeout store 2.
-	waitUntilDead(t, sp, 2)
+	waitUntilDead(t, mc, sp, 2)
 	{
 		sp.mu.RLock()
 		store2, ok := sp.stores[2]
@@ -175,7 +182,7 @@ func TestStorePoolDies(t *testing.T) {
 	}
 
 	// Timeout store 2 again.
-	waitUntilDead(t, sp, 2)
+	waitUntilDead(t, mc, sp, 2)
 	{
 		sp.mu.RLock()
 		store2, ok := sp.stores[2]
@@ -212,7 +219,7 @@ func verifyStoreList(sp *StorePool, requiredAttrs []string, expected []int) erro
 func TestStorePoolGetStoreList(t *testing.T) {
 	defer leaktest.AfterTest(t)
 	// We're going to manually mark stores dead in this test.
-	stopper, g, sp := createTestStorePool(TestTimeUntilStoreDeadOff)
+	stopper, g, _, sp := createTestStorePool(TestTimeUntilStoreDeadOff)
 	defer stopper.Stop()
 	sg := gossiputil.NewStoreGossiper(g)
 	required := []string{"ssd", "dc"}
@@ -266,7 +273,7 @@ func TestStorePoolGetStoreList(t *testing.T) {
 
 	// Mark one store dead.
 	sp.mu.Lock()
-	sp.stores[deadStore.StoreID].markDead(time.Now())
+	sp.stores[deadStore.StoreID].markDead(sp.clock.Now())
 	sp.mu.Unlock()
 
 	if err := verifyStoreList(sp, required, []int{
@@ -279,7 +286,7 @@ func TestStorePoolGetStoreList(t *testing.T) {
 
 func TestStorePoolGetStoreDetails(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	stopper, g, sp := createTestStorePool(TestTimeUntilStoreDeadOff)
+	stopper, g, _, sp := createTestStorePool(TestTimeUntilStoreDeadOff)
 	defer stopper.Stop()
 	sg := gossiputil.NewStoreGossiper(g)
 	sg.GossipStores(uniqueStore, t)
@@ -295,7 +302,7 @@ func TestStorePoolGetStoreDetails(t *testing.T) {
 
 func TestStorePoolFindDeadReplicas(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	stopper, g, sp := createTestStorePool(TestTimeUntilStoreDead)
+	stopper, g, mc, sp := createTestStorePool(TestTimeUntilStoreDead)
 	defer stopper.Stop()
 	sg := gossiputil.NewStoreGossiper(g)
 
@@ -357,7 +364,7 @@ func TestStorePoolFindDeadReplicas(t *testing.T) {
 		t.Fatalf("expected no dead replicas initially, found %d (%v)", len(deadReplicas), deadReplicas)
 	}
 	// Timeout all stores, but specifically store 5.
-	waitUntilDead(t, sp, 5)
+	waitUntilDead(t, mc, sp, 5)
 
 	// Resurrect all stores except for 4 and 5.
 	sg.GossipStores(stores[:3], t)

--- a/storage/store_test.go
+++ b/storage/store_test.go
@@ -133,9 +133,9 @@ func createTestStoreWithoutStart(t *testing.T) (*Store, *hlc.ManualClock, *stop.
 	rpcContext := rpc.NewContext(&base.Context{}, hlc.NewClock(hlc.UnixNano), stopper)
 	ctx := TestStoreContext
 	ctx.Gossip = gossip.New(rpcContext, gossip.TestBootstrap)
-	ctx.StorePool = NewStorePool(ctx.Gossip, TestTimeUntilStoreDeadOff, stopper)
 	manual := hlc.NewManualClock(0)
 	ctx.Clock = hlc.NewClock(manual.UnixNano)
+	ctx.StorePool = NewStorePool(ctx.Gossip, ctx.Clock, TestTimeUntilStoreDeadOff, stopper)
 	eng := engine.NewInMem(roachpb.Attributes{}, 10<<20, stopper)
 	ctx.Transport = multiraft.NewLocalRPCTransport(stopper)
 	stopper.AddCloser(ctx.Transport)

